File: macro pass2/macroPass2.cpp
#include <bits/stdc++.h>
using namespace std;

vector<string> readLines(const string& filename) {
    ifstream file(filename);
    vector<string> lines;
    string line;

    while (getline(file, line)) {
        lines.push_back(line);
    }
    return lines;
}

vector<string> split(const string& str, char delimiter) {
    vector<string> tokens;
    string token;
    stringstream tokenStream(str);

    while (getline(tokenStream, token, delimiter)) {
        tokens.push_back(token);
    }
    return tokens;
}

class Process {
public:
    static string getMacroDef(const string& name, const vector<string>& MNT) {
        for (const auto& entry : MNT) {
            if (split(entry, ' ')[0] == name) {
                return entry; 
            }
        }
        return "";
    }

    static void macroMDT(const string& macroMNT, const vector<string>& MDT, const vector<string>& aptab) {
        int mdtIndex = stoi(split(macroMNT, ' ')[3]) - 1;  
        cout << "Expanded Code:" << endl;

        for (size_t i = mdtIndex; i < MDT.size(); ++i) {
            if (MDT[i] == "MEND") {
                break;  
            } else {
                vector<string> mdtLine = split(MDT[i], ' ');
                string instruction = mdtLine[0];
                string expandedLine = instruction;

                for (size_t j = 1; j < mdtLine.size(); ++j) {
                    string param = mdtLine[j];
                    if (param.substr(0, 3) == "(P,") {
                        int paramIndex = stoi(param.substr(3, param.size() - 4)) - 1;  
                        expandedLine += " " + aptab[paramIndex];  
                    } else {
                        expandedLine += " " + param; 
                    }
                }
                cout << expandedLine << endl;  
            }
        }
        cout << "_________________________________" << endl;
    }

    static vector<string> prepareAPTAB(const string& macroMNT, const vector<string>& KPDTAB, const string& macroCall) {
        vector<string> mntValues = split(macroMNT, ' ');
        int numPP = stoi(mntValues[1]);  
        int numKP = stoi(mntValues[2]); 
        int kpdTabStart = stoi(mntValues[4]) - 1; 

        vector<string> aptab;
        vector<string> callParams = split(split(macroCall, ' ')[1], ','); 

        for (const auto& param : callParams) {
            if (param.find('=') == string::npos) {
                aptab.push_back(param); 
            }
        }

        for (size_t i = kpdTabStart; i < kpdTabStart + numKP; ++i) {
            vector<string> entry = split(KPDTAB[i], ' ');
            aptab.push_back(entry[1]); 
        }

        for (const auto& param : callParams) {
            if (param.find('=') != string::npos) {
                vector<string> kvPair = split(param, '=');
                aptab.push_back(kvPair[1]); 
            }
        }
        return aptab; 
    }
};

int main() {
    vector<string> macroCalls = readLines("macro_call.txt");
    vector<string> MNT = readLines("MNT2.txt");
    vector<string> MDT = readLines("MDT2.txt");
    vector<string> KPDTAB = readLines("KPDTAB2.txt");

    for (auto& macroCall : macroCalls) {
        string macroName = split(macroCall, ' ')[0];  
        string macroEntry = Process::getMacroDef(macroName, MNT);  
        vector<string> aptab = Process::prepareAPTAB(macroEntry, KPDTAB, macroCall);
        Process::macroMDT(macroEntry, MDT, aptab);
    }

    return 0;
}

File: macro2/macro_pass2.py
call_file = open("CALLS2.txt", "r")
lines = call_file.readlines()
macro_calls = [line.strip() for line in lines]

mnt_file = open("MNT.txt", "r")
lines = mnt_file.readlines()
MNT = [line.strip() for line in lines]

mdt_file = open("MDT.txt", "r")
lines = mdt_file.readlines()
MDT = [line.strip() for line in lines]

kpd_file = open("KPTD.txt", "r")
lines = kpd_file.readlines()
KPDTAB = [line.strip() for line in lines]

print(KPDTAB)


class Process:
    @staticmethod
    def get_macro_def(name):
        for entry in MNT:
            # print("printing Entry")
            # print(entry)
            if entry.split()[0] == name:
                return entry

    @staticmethod
    def macro_mdt(macro_mnt, aptab):
        mdt_index = int(macro_mnt.split()[3]) - 1
        # print("Expanded Code")

        for line in MDT[mdt_index:]:
            if line == "MEND":
                break
            else:
                mdt_line = line.split()
                instruction = mdt_line[0]
                expanded_line = instruction

                for param in mdt_line[1:]:
                    if param.startswith("(P,"):
                        param_index = int(param[3:-1]) - 1
                        expanded_line += f" {aptab[param_index]}"
                    else:
                        expanded_line += f" {param}"

                print(expanded_line)
        print("_")

    @staticmethod
    def prepare_aptab(macro_mnt, macro_call):

        # macro_mnt example = INCR_D 0 3 9 4
        mnt_values = macro_mnt.split()
        num_pp, num_kp, kpdtab_start = int(mnt_values[1]), int(mnt_values[2]), int(mnt_values[4]) - 1

        aptab = []
        # macro call example = INCR_M MEM_VAL=A,INCR_VAL=B,REG=AREG
        call_params = macro_call.split()[1].split(",")
        print(macro_call.split()[0])

        # positional parameters added in aptab
        for param in call_params:
            if "=" not in param:
                aptab.append(param)
        # Adding default values for KP from KPDTAB
        kp_defaults = {}

        # entry example = MEM_VAL no_default
        for entry in KPDTAB[kpdtab_start:kpdtab_start + num_kp]:  # kpdtp to kpdtp + kp
            param_name, param_value = entry.split()
            kp_defaults[param_name] = param_value

        # over riding the default entries from macro call
        for param in call_params:
            if "=" in param:
                param_name, param_value = param.split("=")
                kp_defaults[param_name] = param_value

        # Fill in APTAB with positional and keyword parameters
        for key in kp_defaults:
            aptab.append(kp_defaults[key])

        return aptab


class Main:
    @staticmethod
    def main():
        for macro_call in macro_calls:
            macro_name = macro_call.split()[0]
            macro_entry = Process.get_macro_def(macro_name)
            aptab = Process.prepare_aptab(macro_entry, macro_call)
            Process.macro_mdt(macro_entry, aptab)


Main.main()